---
title: "Report-8.21"
author: "chuhan"
date: "2020/08/21"
output:
  html_document: 
    df_print: paged
    toc: yes
    theme: united
    toc_float: yes
    collapsed: yes
    smooth_scroll: yes
  pdf_docuument:
    toc: yes
---

```{r, setup, include=FALSE}
knitr::opts_chunk$set(echo = F)
```


# 空间转录组一般流程

## *质量控制
```{r message=FALSE, warning=FALSE, paged.print=TRUE}
rm(list = ls())
#devtools::install_github('satijalab/seurat-data')
library(Seurat)
library(SeuratData)
library(ggplot2)
library(patchwork)
library(dplyr)
library(SingleR)
library(R.matlab)
library(pheatmap)
library(dplyr)
library(monocle)
library(CellChat)
library(ggalluvial)
library(svglite)
library(NMF)
options(stringsAsFactors = FALSE)
setwd("e://data/work/xinqiao/2020.7/report/")

brain <- Load10X_Spatial("E:/data/work/work/jia/test nishi and kongjian/ruxianai-a1/")
plot1 <- VlnPlot(brain, features = "nCount_Spatial", pt.size = 0.1) + NoLegend()
plot2 <- SpatialFeaturePlot(brain, features = "nCount_Spatial") + theme(legend.position = "right")
wrap_plots(plot1, plot2)
```



## *数据预处理

```{r warning=FALSE, paged.print=TRUE}
brain <- SCTransform(brain, assay = "Spatial", verbose = FALSE)
brain@assays$SCT@data[50:55, 53:55]
```



## *可视化关注的基因
```{r warning=FALSE, paged.print=TRUE}
SpatialFeaturePlot(brain, features = c("HIF1A", "CA9"))
```



## *降维，聚类和可视化

```{r fig.width=12, warning=FALSE, paged.print=TRUE}
brain <- RunPCA(brain, assay = "SCT", verbose = FALSE)
ElbowPlot(brain,ndims = 50)
brain <- FindNeighbors(brain, reduction = "pca", dims = 1:50,verbose = F)
brain <- FindClusters(brain, verbose = FALSE)
brain <- RunUMAP(brain, reduction = "pca", dims = 1:50, verbose = F)
p1 <- DimPlot(brain, reduction = "umap", label = TRUE)
p2 <- SpatialDimPlot(brain, label = TRUE, label.size = 3)
print(p1 + p2)
#SpatialDimPlot(brain, cells.highlight = CellsByIdentities(object = brain, idents = c(1, 2, 5, 3, 4, 8)), facet.highlight = TRUE, ncol = 3)
#SpatialFeaturePlot(brain, features = "HIF1A", interactive = TRUE)
```



## *差异基因(基于基因)

--(cluster 1)展示前六个差异基因的表达
```{r fig.height=15, fig.width=12, warning=FALSE, paged.print=TRUE}
de_markers <- FindMarkers(brain, ident.1 = 1,  verbose = F)
SpatialFeaturePlot(object = brain, features = rownames(de_markers)[1:6], alpha = c(0.1, 1), ncol = 3)

#brain <- FindSpatiallyVariableFeatures(brain, assay = "SCT", features = VariableFeatures(brain)[1:3000], selection.method = "markvariogram")
#top.features <- head(SpatiallyVariableFeatures(brain, selection.method = "markvariogram"), 6)
#SpatialFeaturePlot(brain, features = top.features, ncol = 3, alpha = c(0.1, 1))
```

假如降维后umap结果映射到空间中有明显限制，效果很好，上述方法有效~

当效果不好时，结合空间分布与基因表达，搜索表达出空间特征的特征基因，可探索方法：

1.Identification of spatial expression trends in single-cell gene expression data (10.1038/nmeth.4634)(seurat)

2. SpatialDE: identification of spatially variable genes (10.1038/nmeth.4636)

3. Splotch: Robust estimation of aligned spatial temporal gene expression data (10.1101/757096)

4. Giotto (http://spatialgiotto.rc.fas.harvard.edu/) 


# 空间转录组数据整合

## *空间转录组整合单细胞转录组

可用于预测空间细胞类型

--对空间转录组的每个spot有一个关于单细胞群落的评分矩阵
```{r message=FALSE, warning=FALSE, paged.print=TRUE}
if (!file.exists("ST&SC.Rdata")) {
#InstallData("stxBrain")
brain <- LoadData("stxBrain", type = "anterior1")
brain <- SCTransform(brain, assay = "Spatial", verbose = FALSE)
brain <- RunPCA(brain, assay = "SCT", verbose = FALSE)
brain <- FindNeighbors(brain, reduction = "pca", dims = 1:30, verbose = FALSE)
brain <- FindClusters(brain, verbose = FALSE)
brain <- RunUMAP(brain, reduction = "pca", dims = 1:30, verbose = FALSE)
cortex <- subset(brain, idents = c(1, 2, 3, 5, 6, 7))
# now remove additional cells, use SpatialDimPlots to visualize what to remove
# SpatialDimPlot(cortex,cells.highlight = WhichCells(cortex, expression = image_imagerow > 400 |
# image_imagecol < 150))
cortex <- subset(cortex, anterior1_imagerow > 400 | anterior1_imagecol < 150, invert = TRUE)
cortex <- subset(cortex, anterior1_imagerow > 275 & anterior1_imagecol > 370, invert = TRUE)
cortex <- subset(cortex, anterior1_imagerow > 250 & anterior1_imagecol > 440, invert = TRUE)
allen_reference <- readRDS("e://data/work/xinqiao/2020.7/report/allen_cortex.rds")
allen_reference <- SCTransform(allen_reference, ncells = 3000, verbose = FALSE) %>% RunPCA(verbose = FALSE) %>% RunUMAP(dims = 1:30,verbose = FALSE)
cortex <- SCTransform(cortex, assay = "Spatial", verbose = FALSE) %>% RunPCA(verbose = FALSE)
# the annotation is stored in the 'subclass' column of object metadata
#DimPlot(allen_reference, group.by = "subclass", label = TRUE)
save(cortex,allen_reference, file = "ST&SC.Rdata")
} else {
  load("ST&SC.Rdata")
}
anchors <- FindTransferAnchors(reference = allen_reference, query = cortex, normalization.method = "SCT",verbose = F)
predictions.assay <- TransferData(anchorset = anchors, refdata = allen_reference$subclass, prediction.assay = TRUE, weight.reduction = cortex[["pca"]], verbose = F )
cortex[["predictions"]] <- predictions.assay
DefaultAssay(cortex) <- "predictions"
SpatialFeaturePlot(cortex, features = c("L2/3 IT", "L4", "L6 IT", "Oligo"), pt.size.factor = 1.6, ncol = 2, crop = TRUE)
```

--利用上述算法一，把评分矩阵当作表达，搜索表达出空间特征的评分集，按排名选取展示
```{r fig.height=15, fig.width=18, warning=FALSE, paged.print=TRUE}
cortex <- FindSpatiallyVariableFeatures(cortex, assay = "predictions", selection.method = "markvariogram", features = rownames(cortex), r.metric = 5, slot = "data",verbose = F)
top.clusters <- head(SpatiallyVariableFeatures(cortex), 4)
SpatialPlot(object = cortex, features = top.clusters, ncol = 2)
```

## *利用单细胞对空间转录组细胞比例分解

--已存在的方法MIA，Giotto，CLBERSORT，SPOTlight，bisqueRNA

重点关注和重现了MIA，算法来源：

Integrating microarray-based spatial transcriptomics and single-cell RNA-seq reveals tissue architecture in pancreatic ductal adenocarcinomas https://doi.org/10.1038/s41587-019-0392-8
```{r warning=FALSE, paged.print=TRUE}
mat <- readMat("e://data/work/xinqiao/2020.7/forth/MIA/Multimodal-intersection-analysis-MIA--master/Multimodal-intersection-analysis-MIA--master/MIA_example.mat")
ST <- mat[["ST.region.genes"]]
ST <- as.data.frame(ST)
colnames(ST) <- unlist(mat[["tissue.region.labels"]])             
rownames(ST) <- unlist(mat[["gene.names"]])

SC <- mat[["cell.type.genes"]]
SC <- as.data.frame(SC)
colnames(SC) <- unlist(mat[["cell.type.labels"]])
rownames(SC) <- unlist(mat[["gene.names"]])

M <- matrix(ncol = ncol(SC),nrow = ncol(ST))
M_enr <- matrix(ncol = ncol(SC),nrow = ncol(ST))
M_dep <- matrix(ncol = ncol(SC),nrow = ncol(ST))
geneList <- rownames(ST)
b <- length(geneList)

for (region in 1:ncol(ST)) {
  #region=1
  G <- rownames(ST[which(ST[,region]==1),])
  for (celltype in 1:ncol(SC)) {
     #celltype=2
     C <-  rownames(SC[which(SC[,celltype]==1),])
     a <- length(intersect(G,C))
    c <- length(G)
    d <- length(C)
    M_enr[region,celltype] <-  -1*log10(phyper(c-a-1,c,b-c,b-d))
    M[region,celltype] <- M_enr[region,celltype]
      M_dep[region,celltype] <- -1*log10(1-(phyper(c-a-1,c,b-c,b-d)))
     if (M[region,celltype] < M_dep[region,celltype]) {
         M[region,celltype] <-  -1*M_dep[region,celltype]
      }
    
    }  
}

M <- t(M)
rownames(M) <- colnames(SC)
colnames(M) <- colnames(ST)
M[M > 10] = 10
M[M < -10] = -10
pheatmap::pheatmap(M,scale = "none",cluster_cols = F,cluster_rows = F)
```



## *空间多个切片的整合

--连续或相近切片
```{r fig.width=15, message=FALSE, warning=FALSE, paged.print=TRUE}
rm(list = ls())
if (!file.exists("brain_merge.Rdata")) {
brain <- Load10X_Spatial("E:/data/work/work/jia/test nishi and kongjian/ruxianai-a1/", )
brain[["orig.ident"]] <- "Breast1"
brain <- SCTransform(brain, assay = "Spatial", verbose = FALSE)
brain <- RunPCA(brain, assay = "SCT", verbose = FALSE)
brain <- FindNeighbors(brain, reduction = "pca", dims = 1:50,verbose = F)
brain <- FindClusters(brain, verbose = FALSE)
brain <- RunUMAP(brain, reduction = "pca", dims = 1:50, verbose = F)

brain2 <- Load10X_Spatial("E:/data/work/work/jia/test nishi and kongjian/ruxianai-a2/")
brain2[["orig.ident"]] <- "Breast2"
brain2 <- SCTransform(brain2, assay = "Spatial", verbose = FALSE)
brain.merge <- merge(brain, brain2)

DefaultAssay(brain.merge) <- "SCT"
VariableFeatures(brain.merge) <- c(VariableFeatures(brain), VariableFeatures(brain2))
brain.merge <- RunPCA(brain.merge, verbose = FALSE)
brain.merge <- FindNeighbors(brain.merge, dims = 1:30,verbose = FALSE)
brain.merge <- FindClusters(brain.merge, verbose = FALSE)
brain.merge <- RunUMAP(brain.merge, dims = 1:30,verbose = FALSE)
save(brain,brain2,brain.merge, file = "brain_merge.Rdata")
} else {
  load("brain_merge.Rdata")
}
DimPlot(brain.merge, reduction = "umap", group.by = c("ident", "orig.ident"))
```
```{r fig.height=15, fig.width=18, message=FALSE, warning=FALSE, paged.print=TRUE}
SpatialDimPlot(brain.merge)
SpatialFeaturePlot(brain.merge, features = c("HIF1A", "CA9"))
```



# 空间转录组可探索的分析点

## *时序分析
```{r fig.width=12, warning=FALSE, paged.print=TRUE}
if (!file.exists("HSMM_myo.Rdata")) {
data <- brain@assays$Spatial@counts
new_pd <- brain@meta.data
pd <- new('AnnotatedDataFrame', data = brain@meta.data)
fData <- data.frame(gene_short_name = row.names(data), row.names = row.names(data))
fd <- new('AnnotatedDataFrame', data = fData)
HSMM <- newCellDataSet(data,
                phenoData = pd,
                featureData = fd,
                expressionFamily=negbinomial.size())
HSMM <- estimateSizeFactors(HSMM)
HSMM <- estimateDispersions(HSMM)
HSMM <- detectGenes(HSMM, min_expr = 0.1)
expressed_genes <- row.names(subset(fData(HSMM),num_cells_expressed >= 10))
HSMM_myo <- HSMM
diff_test_res <- differentialGeneTest(HSMM_myo[expressed_genes,],fullModelFormulaStr = "~seurat_clusters")
ordering_genes <- row.names (subset(diff_test_res, qval < 0.01))
HSMM_myo <- setOrderingFilter(HSMM_myo, ordering_genes)
plot_ordering_genes(HSMM_myo)
HSMM_myo <- reduceDimension(HSMM_myo, max_components = 2,method = 'DDRTree')
HSMM_myo <- orderCells(HSMM_myo)
save(HSMM_myo, file = "HSMM_myo.Rdata")
} else {
  load("HSMM_myo.Rdata")
}
p1 <- plot_cell_trajectory(HSMM_myo, color_by = "Pseudotime")
brain@meta.data$Pseudotime <- HSMM_myo@phenoData@data$Pseudotime
p2 <- SpatialFeaturePlot(brain, features = "Pseudotime") + theme(legend.position = "right")
print(p1 + p2)
```



## *细胞互作

```{r warning=FALSE, paged.print=TRUE}
rm(list = ls())
if (!file.exists("cell_inte_brain.Rdata")) {
brain <- Load10X_Spatial("E:/data/work/work/jia/zhongliuhelinba/linba/outs/")
brain <- SCTransform(brain, assay = "Spatial", verbose = FALSE)
brain <- RunPCA(brain, assay = "SCT", verbose = FALSE)
brain <- FindNeighbors(brain, reduction = "pca", dims = 1:50,verbose = F)
brain <- FindClusters(brain, verbose = FALSE)
brain <- RunUMAP(brain, reduction = "pca", dims = 1:50, verbose = F)

singleR_input <- GetAssayData(brain)
hpca.se_sec <- DatabaseImmuneCellExpressionData()
pred.hesc_sec <- SingleR(test = singleR_input, ref = hpca.se_sec, labels = hpca.se_sec$label.main)
brain@meta.data$singleR_labels_sec <- pred.hesc_sec$labels
save(brain, file = "cell_inte_brain.Rdata")
} else {
  load("cell_inte_brain.Rdata")
}

p1 <- DimPlot(brain, reduction = "umap",label=TRUE,label.size=3)
#p2 <- DimPlot(brain, reduction = "umap",label=TRUE,label.size=3,group.by ="singleR_labels_sec")
#print(p1+p2)
p3 <- SpatialDimPlot(brain, label = TRUE, label.size = 3)
#p4 <- SpatialDimPlot(brain, label = TRUE, label.size = 3,,group.by ="singleR_labels_sec")
print(p1+p3)

if (!file.exists("cell_inte_cellchat.Rdata")) {
data.input <- GetAssayData(brain, assay = "SCT", slot = "data") # normalized data matrix
labels <- brain@meta.data$seurat_clusters
identity <- data.frame(group = labels, row.names = rownames( brain@meta.data))
cellchat <- createCellChat(data = data.input)
cellchat <- addMeta(cellchat, meta = identity, meta.name = "labels")
cellchat <- setIdent(cellchat, ident.use = "labels") # set "labels" as default cell identity
#levels(cellchat@idents)
groupSize <- as.numeric(table(cellchat@idents))

CellChatDB <- CellChatDB.human # use CellChatDB.human if running on human data
CellChatDB.use <- subsetDB(CellChatDB, search = "Secreted Signaling") # use Secreted Signaling for cell-cell communication analysis
cellchat@DB <- CellChatDB.use # set the used database in the object

cellchat <- subsetData(cellchat) # subset the expression data of signaling genes for saving computation cost
future::plan("multiprocess", workers = 4) # do parallel
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)
cellchat <- projectData(cellchat, PPI.human)

cellchat <- computeCommunProb(cellchat)
cellchat <- computeCommunProbPathway(cellchat)
cellchat <- aggregateNet(cellchat)
save(cellchat, groupSize, file = "cell_inte_cellchat.Rdata")
} else {
  load("cell_inte_cellchat.Rdata")
}

#levels(cellchat@idents)
#head(cellchat@LR$LRsig)
#cellchat@netP$pathways
vertex.receiver = seq(1,7) # a numeric vector
pathways.show <- "TGFb"

netVisual_aggregate(cellchat, signaling = pathways.show,  vertex.receiver = vertex.receiver, vertex.size = groupSize)
```
```{r}
netVisual_aggregate(cellchat, signaling = c("TGFb"), layout = "circle", vertex.size = groupSize)
```

```{r}
netAnalysis_contribution(cellchat, signaling = pathways.show)
```

```{r}
cellchat <- netAnalysis_signalingRole(cellchat, slot.name = "netP") # the slot 'netP' means the inferred intercellular communication network of signaling pathways
netVisual_signalingRole(cellchat, signaling = pathways.show, width = 12, height = 2.5, font.size = 10)
```

```{r }
nPatterns = 7
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "outgoing", k = nPatterns)
# Visualize the communication pattern using river plot
netAnalysis_river(cellchat, pattern = "outgoing",font.size = 2.5)
# Visualize the communication pattern using dot plot
netAnalysis_dot(cellchat, pattern = "outgoing",font.size =8)

#myidentifyCommunicationPatterns <- edit(identifyCommunicationPatterns)
#environment(myidentifyCommunicationPatterns) <- environment(identifyCommunicationPatterns)
#cellchat <- myidentifyCommunicationPatterns(cellchat, pattern = "outgoing", k = nPatterns)
```

```{r }
nPatterns = 5 
cellchat <- identifyCommunicationPatterns(cellchat, pattern = "incoming", k = nPatterns)
netAnalysis_river(cellchat, pattern = "incoming",font.size = 2.5)
netAnalysis_dot(cellchat, pattern = "incoming",font.size =8)
```